### PR TITLE
requires_grad attribute for meta tensors #96

### PR DESCRIPTION
**Pull request by kovalevfm**
[ #96]

### CODE REVIEW

File: test/test_tensordict.py [hunk 1]
```diff
@@ -1438,6 +1438,37 @@ def test_getitem_batch_size():
...
+                    11, 12, 5, device=device, requires_grad=True if i == 5 else False
+                ),
+                "key2": torch.zeros(
+                    11, 12, 50, device=device, dtype=torch.bool
+                ).bernoulli_(),
+            },
+        )
+        for i in range(10)
+    ]
+    sub_td = LazyStackedTensorDict(*tensordicts, stack_dim=0)
```
Comment (vmoens) [inline test/test_tensordict.py:1459]: maybe `stacked_td` instead of `sub_td`?
You could try to index the tensordict too by the way:
```python
sub_td1 = td[0]
assert sub_td1.get("key1").requires_grad
assert sub_td1.get_meta("key1").requires_grad

sub_td2 = td[torch.tensor([1, 2])]
assert sub_td2.get("key1").requires_grad
assert sub_td2.get_meta("key1").requires_grad
```

We could also do the same for unsqueeze and/or view?

File: test/test_tensordict.py [hunk 2]
```diff
@@ -1438,6 +1438,37 @@ def test_getitem_batch_size():
         assert expected_shape == resulting_shape, idx
 
 
+@pytest.mark.parametrize("device", get_available_devices())
+def test_requires_grad(device):
```
Comment (vmoens) [inline test/test_tensordict.py:1442]: This looks great to test that tensordict support gradients when created, but what about modification of existing tensordicts?


Do you think it would be possible to also test this with `TestTensorDicts`? 
You can go through various examples of TensorDicts with 
```python
    @pytest.mark.parametrize(
        "td_name",
        [
            "td",
            "stacked_td",
            "sub_td",
            "idx_td",
            "saved_td",
            "unsqueezed_td",
            "td_reset_bs",
        ],
    )
    def test_blabla(td_name):
        exec()
```

What you would need to do is replace one tensor of the tensordict with a tensor that requires a gradient and check that this is supported. Something like
```python
td.set("a", torch.randn_like(td.get("a")).requires_grad_())
```
You'd need to check that the MetaTensor associated with that tensor is updated too.
```python
assert td.get_meta("a").requires_grad
```

**Review by vmoens — CHANGES_REQUESTED** [review]

File: test/test_tensordict.py (hunk 1 above)
Comment (kovalevfm) [reply test/test_tensordict.py]: I've added test class with different TensorDicts with requires_grad and some operations like view and squeeze.

File: test/test_tensordict.py [hunk 3]
```diff
@@ -1115,6 +1115,159 @@ def test_items_values_keys(self, td_name):
...
+            "idx_td",
+            "saved_td",
+            "unsqueezed_td",
+            "td_reset_bs",
+        ],
+    )
+    def test_set_requires_grad(self, td_name):
+        td = getattr(self, td_name)
+        assert not td._get_meta("a").requires_grad
+        td.set("a", torch.randn_like(td.get("a")).requires_grad_())
```
Comment (vmoens) [inline test/test_tensordict.py:1133]: we probably want to update that given the new API of `sub_td.set`

File: test/test_tensordict.py [hunk 4]
```diff
@@ -1115,6 +1115,159 @@ def test_items_values_keys(self, td_name):
...
+        assert td._get_meta("a").requires_grad
+
+
+class TestTensorDictsRequiresGrad:
```
Comment (vmoens) [inline test/test_tensordict.py:1137]: oh wow that's a whole new class!
I'm ok with it, great work

**Review by vmoens — APPROVED** [review]
Let's just make sure it runs with the new API but then we're good to go

File: test/test_tensordict.py (hunk 4 above)
Comment (kovalevfm) [reply test/test_tensordict.py:1140]: It helped to fund couple bugs, totally worth it

Comment (vmoens) [reply test/test_tensordict.py:1140]: indeed!

**Review by vmoens — APPROVED** [review]
